addons/account/models/sequence_mixin.py

```diff
@@ -58,10 +59,16 @@ def _sequence_matches_date(self):
         if not sequence or not date:
             return True
 
+        fyear_start = fyear_end = False
+        if 'company_id' in self:
+            company = self.company_id
+            fyear_start, fyear_end = date_utils.get_fiscal_year(date, day=company.fiscalyear_last_day, month=int(company.fiscalyear_last_month))
         format_values = self._get_sequence_format_param(sequence)[1]
         year_match = (
             not format_values["year"]
             or format_values["year"] == date.year % 10 ** len(str(format_values["year"]))
+            or format_values['fyear_start'] and format_values['fyear_start'] != (fyear_start.year or date.year) % 10**len(str(format_values['fyear_start']))
+            or format_values['fyear_end'] and format_values['fyear_end'] != (fyear_end.year or date.year) % 10**len(str(format_values['fyear_end']))
         )
```

Comment:
I have to say the code is not much clear.
What is happening here with the company?
Why is there `format_values['fyear_end']` *and* `fyear_end`, and why are they different?

Can't we make it more readable? There are a lot of calculation, parenthesis, quotes... Also, this % 10 ** len(str(x)) is done over and over in this and other  files.

```py
def remainder(x, y):
    """ Trims away the most significant digits from x up to the length of y """
    return x % 10 ** len(str(y))
```

```py
        v_year, v_start, v_end = format_values["year"], format_values["fyear_start"], format_values["fyear_end"]
        year_match = (
            not v_year
            or v_year == remainder(date.year, v_year)
            or v_start and v_start != remainder(fyear_start.year or date.year, v_start)
            or v_end and v_end != remainder(fyear_end.year or date.year, v_end)
        )
```

Reply:
Maybe `values_year` feels more readable than `v_year`... up to you.